In [ ]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.model_selection import KFold, cross_val_score
import numpy as np
from sklearn.preprocessing import scale
from sklearn.datasets import fetch_openml
from sklearn.neighbors import KNeighborsRegressor

Задача 1. 

In [4]:
boston = fetch_openml(name="boston", version=1, as_frame=True)

X = boston.data
y = boston.target

print(X.head())
print("\nцелевая переменная:")
print(y.head())

      CRIM    ZN  INDUS CHAS    NOX     RM   AGE     DIS RAD    TAX  PTRATIO  \
0  0.00632  18.0   2.31    0  0.538  6.575  65.2  4.0900   1  296.0     15.3   
1  0.02731   0.0   7.07    0  0.469  6.421  78.9  4.9671   2  242.0     17.8   
2  0.02729   0.0   7.07    0  0.469  7.185  61.1  4.9671   2  242.0     17.8   
3  0.03237   0.0   2.18    0  0.458  6.998  45.8  6.0622   3  222.0     18.7   
4  0.06905   0.0   2.18    0  0.458  7.147  54.2  6.0622   3  222.0     18.7   

        B  LSTAT  
0  396.90   4.98  
1  396.90   9.14  
2  392.83   4.03  
3  394.63   2.94  
4  396.90   5.33  

целевая переменная:
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: MEDV, dtype: float64


Задача 2

In [7]:
X_scaled = scale(X)

print("Нормализованные признаки:")
print(X_scaled[:5])
print("\nЦелевая переменная:")
print(y.head())

Нормализованные признаки:
[[-0.41978194  0.28482986 -1.2879095  -0.27259857 -0.14421743  0.41367189
  -0.12001342  0.1402136  -0.98284286 -0.66660821 -1.45900038  0.44105193
  -1.0755623 ]
 [-0.41733926 -0.48772236 -0.59338101 -0.27259857 -0.74026221  0.19427445
   0.36716642  0.55715988 -0.8678825  -0.98732948 -0.30309415  0.44105193
  -0.49243937]
 [-0.41734159 -0.48772236 -0.59338101 -0.27259857 -0.74026221  1.28271368
  -0.26581176  0.55715988 -0.8678825  -0.98732948 -0.30309415  0.39642699
  -1.2087274 ]
 [-0.41675042 -0.48772236 -1.30687771 -0.27259857 -0.83528384  1.01630251
  -0.80988851  1.07773662 -0.75292215 -1.10611514  0.1130321   0.41616284
  -1.36151682]
 [-0.41248185 -0.48772236 -1.30687771 -0.27259857 -0.83528384  1.22857665
  -0.51117971  1.07773662 -0.75292215 -1.10611514  0.1130321   0.44105193
  -1.02650148]]

Целевая переменная:
0    24.0
1    21.6
2    34.7
3    33.4
4    36.2
Name: MEDV, dtype: float64


Задача 3

In [10]:
p_values = np.linspace(1, 10, 200)

mean_scores = []

kf = KFold(n_splits=5, shuffle=True, random_state=42)

for p in p_values:
    model = KNeighborsRegressor(n_neighbors=5, weights='distance', metric='minkowski', p=p)
    scores = cross_val_score(model, X_scaled, y, cv=kf, scoring='neg_mean_squared_error')
    mean_scores.append(-scores.mean())

best_p_index = np.argmin(mean_scores)
best_p = p_values[best_p_index]
best_mse = mean_scores[best_p_index]

print(f"оптимальное знач-е p: {best_p}")
print(f"МИН значчение средней квадратической ошибки (MSE): {best_mse}")

оптимальное знач-е p: 1.0
МИН значчение средней квадратической ошибки (MSE): 16.030646734221644


Задача 4

Для определения оптимального значения  **p** в модели kNN я провела поиск по диапазону значений от 1 до 10 с шагом 0.05. Где для  каждого значения p была вычислена средняя ошибка (MSE) на 5-кратной кросс-валидации. 

Минимальное значение средней ошибки (16.03) было достигнуто при p=1.0.